src/db/impl/abstract.ts

```diff
@@ -0,0 +1,255 @@
+    const buf = await this.get(encodeKey(Bucket.chainInfo, Key.finalizedBlock));
+    return deserialize(buf, BeaconBlock).deserializedData;
+  }
+
+  public async setFinalizedBlock(block: BeaconBlock): Promise<void> {
+    await this.put(encodeKey(Bucket.chainInfo, Key.finalizedBlock), serialize(block, BeaconBlock));
+  }
+
+  public async getJustifiedBlock(): Promise<BeaconBlock> {
+    const buf = await this.get(encodeKey(Bucket.chainInfo, Key.justifiedBlock));
+    return deserialize(buf, BeaconBlock).deserializedData;
+  }
+
+  public async setJustifiedBlock(block: BeaconBlock): Promise<void> {
+    await this.put(encodeKey(Bucket.chainInfo, Key.justifiedBlock), serialize(block, BeaconBlock));
+  }
+
+  public async getChainHead(): Promise<BeaconBlock> {
+    const heightBuf = await this.get(encodeKey(Bucket.chainInfo, Key.chainHeight));
+    const height = deserialize(heightBuf, uint64).deserializedData;
+    const blockRoot = await this.get(encodeKey(Bucket.mainChain, height));
+    return await this.getBlock(blockRoot);
+  }
+
+  public async getChainHeadRoot(): Promise<bytes32> {
+    const block = await this.getChainHead();
+    return treeHash(block, BeaconBlock);
+  }
+
+  public async setChainHead(state: BeaconState, block: BeaconBlock): Promise<void> {
+    const blockRoot = treeHash(block, BeaconBlock);
+    const slot = block.slot;
+    // block should already be set
+    await this.getBlock(blockRoot);
+    await this.batchPut([
+      {
+        key: encodeKey(Bucket.mainChain, slot),
+        value: blockRoot
+      },
+      {
+        key: encodeKey(Bucket.chainInfo, Key.chainHeight),
+        value: serialize(slot, uint64)
+      },
+      {
+        key: encodeKey(Bucket.chainInfo, Key.state),
+        value: serialize(state, BeaconState)
+      }
+    ]);
+  }
+
+  public async getAttestations(): Promise<Attestation[]> {
+    const data = await this.search({
+      gt: encodeKey(Bucket.attestation, Buffer.alloc(0)),
+      lt: encodeKey(Bucket.attestation + 1, Buffer.alloc(0)),
+    });
+    return data.map((data) => deserialize(data, Attestation).deserializedData);
+  }
+
+  public async setAttestation(attestation: Attestation): Promise<void> {
+    const attestationRoot = treeHash(attestation, Attestation);
+    await this.put(encodeKey(Bucket.attestation, attestationRoot), serialize(attestation, Attestation));
+  }
+
+  public async deleteAttestations(attestations: Attestation[]): Promise<void> {
+    const criteria: any[] = [];
+    attestations.forEach((n) =>
+      criteria.push(encodeKey(Bucket.attestation, treeHash(n, Attestation)))
+    );
+    await this.batchDelete(criteria);
+  }
+
+  public async getVoluntaryExits(): Promise<VoluntaryExit[]> {
+    const data = this.search({
+      gt: encodeKey(Bucket.exit, Buffer.alloc(0)),
+      lt: encodeKey(Bucket.exit + 1, Buffer.alloc(0)),
+    });
+    return deserialize(data, VoluntaryExit).deserializedData;
```

Comment:
We'll need to loop through the data and deserialize each one, like `getAttestations` above

Reply:
nice catch, not sure how I've missed it